core/controllers/feedback_test.py

```diff
@@ -336,3 +339,109 @@ def _get_email(n):
+                'exploration_version': 2,
+                'state_name': u'state_one',
+                'suggestion_content': {'old_content': {}},
+            }, csrf_token)
+        self.post_json(
+            '%s/%s' % (feconf.SUGGESTION_URL_PREFIX, self.EXP_ID), {
+                'exploration_version': 2,
+                'state_name': u'state_one',
+                'suggestion_content': {'old_content': {}},
+            }, csrf_token)
+
+        # Create a thread without suggestions.
+        self.post_json( 
+            '%s/%s' % (feconf.FEEDBACK_THREADLIST_URL_PREFIX, self.EXP_ID), {
+                'state_name': None,
+                'subject': u'New Thread ¡unicode!',
+                'text': 'Message 0'}, csrf_token)     
+
+        # Get a list of open threads without suggestions. 
+        response_dict = self.get_json( 
+            '%s/%s/%s/%s' % ( 
+                feconf.SUGGESTION_LIST_URL_PREFIX, self.EXP_ID, 'OPEN', False))  
+        self.assertEqual(len(response_dict['threads']), 1)       
+ 
+        # Get a list of all threads with suggestion. 
+        response_dict = self.get_json(
+            '%s/%s/%s/%s' % (
+            feconf.SUGGESTION_LIST_URL_PREFIX, self.EXP_ID, 'ALL', 'True'))
+        threads = response_dict['threads']
+        self.assertEqual(len(threads), 6)
+
+        # Get a suggestion.
+        thread_id_1 = threads[0]['thread_id']
+        thread_id_2 = threads[1]['thread_id']
+        response_dict = self.get_json(
+            '%s/%s/%s' % (feconf.FEEDBACK_THREAD_URL_PREFIX,
+                          self.EXP_ID, thread_id_1.split('.')[1]))
+        self.assertEqual(
+            response_dict['suggestion']['id'], thread_id_1)
+
+        # Accept a suggestion.
+        with self.swap(exp_services, '_is_suggestion_valid',
+                       self._return_true):
+            with self.swap(exp_services, 'update_exploration',
+                           self._return_null):
+                self.put_json(
+                    '%s/%s/%s' % (feconf.SUGGESTION_ACTION_URL_PREFIX,
+                                  self.EXP_ID, thread_id_1.split('.')[1]), {
+                        'action': u'accept',
+                        'change_list': [{
+                            'cmd': 'add_state',
+                            'state_name': 'name'
+                        }],
+                        'commit_message': 'Add new state'},
+                    csrf_token)
+
+        # Reject a suggestion.
+        self.put_json(
+            '%s/%s/%s' % (feconf.SUGGESTION_ACTION_URL_PREFIX,
+                          self.EXP_ID, thread_id_2.split('.')[1]), 
+            {'action': u'reject'}, csrf_token)
+
+        # Get a list of closed threads with suggestion. 
+        response_dict = self.get_json(
+            '%s/%s/%s/%s' % (
+            feconf.SUGGESTION_LIST_URL_PREFIX, self.EXP_ID, 'CLOSED', True))
+        self.assertEqual(len(response_dict['threads']), 2)
```

Comment:
Can you add tests that simulate the following scenarios (and maybe come up with a few more):

- User submits suggestion, Editor rejects it, Editor tries to accept it and an error occurs
- User submits suggestion, Editor rejects it, Different Editor tries to accept it and an error occurs
- User submits suggestion, Editor accepts it, Editor tries to reject it and an error occurs
- User submits suggestion, Editor accepts it, Different Editor tries to reject it and an error occurs
- User submits suggestion, User tries to accept it and an error occurs (insufficient rights)
- User submits suggestion, Editor makes a change to the exploration that causes a merge conflict, Editor tries to accept suggestion and an error occurs due to merge conflicts
- User submits suggestion, Editor makes a change to the exploration that does not cause a merge conflict, Editor accepts suggestion
- Editor tries to accept/reject suggestion without specifying certain query parameters (400 error)

Basically try to trace through the different flows in the flow diagram in the design doc, and make sure they all produce correct behaviour (including correct error handling).